File: problem-3/palindrome.py
# python solution for finding the longest palindromic substring
# expand around center includes checking iterating characters deemec as the centre, expanding around them and saving the longest palindrome found
# in the case of even length palindromes, we deem the center to be between two characters
# this solution has a time complexity of O(n^2), better thn the brute force O(n^3) solution

class Solution:
    def longestPalindrome(self, s: str) -> str:
        result = ""
        result_length = 0

        for i in range(len(s)):
            # odd length palindromes
            left, right = i, i
            while left >= 0 and right < len(s) and s[left] == s[right]:
                if (right - left + 1) > result_length:
                    result = s[left:right + 1]
                    result_length = right - left + 1
                left -= 1
                right += 1

            # even length palindromes
            left, right = i, i + 1
            while left >= 0 and right < len(s) and s[left] == s[right]:
                if (right - left + 1) > result_length:
                    result = s[left:right + 1]
                    result_length = right - left + 1
                left -= 1
                right += 1

        return result
                    

File: problem-3/Palindrome.java
// longest palindromic substring in a string solution but in java

public class Palindrome{

    private static int start = 0;
    private static int maxLength = 0;

    public static String longestPalindrome_ExpandAroundCenter(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }
        
        start = 0;
        maxLength = 0;

        for (int i = 0; i < s.length(); i++) {
            // odd length palindromes (center is at i)
            expandAroundCenter(s, i, i);
            
            // even length palindromes (center is between i and i+1)
            expandAroundCenter(s, i, i + 1);
        }

        return s.substring(start, start + maxLength);
    }

    private static void expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
       
        int currentLength = right - left - 1;
        if (currentLength > maxLength) {
            maxLength = currentLength;
            start = left + 1;
        }
    }

    public static void main(String[] args) {
        String test = "banana";
        System.out.println("Expand Around Center Result: " + longestPalindrome_ExpandAroundCenter(test));
    }
}